llvm/lib/Target/AIE/AIERegClassConstrainer.cpp

```diff
@@ -0,0 +1,135 @@
+//===-- AIERegClassConstrainer.cpp - Constrain Reg classes before RA ------===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/CodeGen/LiveIntervals.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineRegisterInfo.h"
+#include "llvm/CodeGen/SlotIndexes.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include <map>
+
+#include "AIE.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "aie-ra"
+
+namespace {
+
+/// Constrain partially allocatable classes to fully allocatable ones.
+/// This makes sure that greedy works with classes for which it is easy to
+/// generate spill code.
+class AIERegClassConstrainer : public MachineFunctionPass {
+
+public:
+  static char ID;
+  AIERegClassConstrainer() : MachineFunctionPass(ID) {}
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    AU.addPreserved<SlotIndexes>();
+    AU.addPreserved<LiveIntervals>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  bool runOnMachineFunction(MachineFunction &Fn) override;
+
+private:
+  /// Find the largest subclass of \p RC that is suitable for register
+  /// allocation. That is, it does not contain reserved registers, as those
+  /// are harder to spill/reload.
+  const TargetRegisterClass &getSuitableRCForRA(const TargetRegisterClass &RC,
+                                                const MachineRegisterInfo &MRI);
+
+  /// Cache for \p getSuitableRCForRA
+  std::map<unsigned, const TargetRegisterClass *> SuitableRCForRA;
```

Comment:
I agree, in the general case it is faster. But I think for small maps it doesn't really have an impact, and it might sometimes even be faster because there's no need for calling a hashing function.
I'm also generally wary of un-ordered data structures and avoid them if possible because they can lead to indeterministic behaviour (not here though), or just complicate debugging. Again, this doesn't really have an impact here, so I would keep it if you're fine with it :)